### PR TITLE
UpdateProgress can attach to particular concurrency queues

### PR DESCRIPTION
**Pull request by tomasherceg**
The `UpdateProgress` control was by default attached to every postback in any concurrency queue. 
This behavior causes issues with some kinds of postbacks, e.g. `LoadItems` event in some Business Pack controls. The UpdateProgress will appear while the user types in the textbox because there is a postback.

This PR adds `IncludedQueues` and `ExcludedQueues` properties to the `UpdateProgress` control. The default behavior is the same as it was before - the control responds to all postbacks from any concurrency queue. But if either of the properties is set, the control will respond only to the included queues, or to any except for the excluded queues respectively.

Additionally, some controls in Business Pack need to use different concurrency queue for different properties (`LoadItems` should run in some background queue that could be easily excluded from the `UpdateProgress` while the `Changed` event should go to the default queue and the `UpdateProgress` should indicate the postback). That's why I've also added the `PostBack.ConcurrencyQueueSettings` collection that can specify a different queue for different events. The `PostBack.ConcurrencyQueue` property remains as is - it can just be overriden in this new collection.

I assume that the Business Pack controls can insert a default entry into this collection for the `LoadItems` property - then the user will be able to just add something like `ExcludedQueues="autocompletion"` and the `UpdateProgress` will ignore these events. And of course everyone can change this default behavior as the collection can be modified from the markup or using the server-side styles.

### CODE REVIEW

File: src/DotVVM.Framework/Resources/Scripts/DotVVM.ts [hunk 1]
```diff
@@ -1246,8 +1255,21 @@ class DotVVM {
...
+
+                    if (includedQueues.length === 0) {
+                        for (let queue in dotvvm.postbackQueues) {
+                            if (excludedQueues.indexOf(queue) < 0 && dotvvm.postbackQueues[queue].noRunning) {
+                                shouldRun = true;
+                                break;
+                            }
+                        }
+                    } else {
+                        shouldRun = includedQueues.some(q => dotvvm.postbackQueues[q] && dotvvm.postbackQueues[q].noRunning);
```
Comment (exyi) [inline src/DotVVM.Framework/Resources/Scripts/DotVVM.ts]: This condition (and the complementary one above too) seems strange, I think the `noRunning` means something different from "nothing is running in this queue". However, I can't look for context on my phone, so maybe I just don't get something... will have a look at that later.

File: src/DotVVM.Framework/Controls/ConcurrencyQueueSettingsCollection.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿using System.Collections.Generic;
+
+namespace DotVVM.Framework.Controls
+{
+    public class ConcurrencyQueueSettingsCollection : List<ConcurrencyQueueSetting>
```
[the remark this answers is not in the document]
Comment (tomasherceg) [reply src/DotVVM.Framework/Controls/ConcurrencyQueueSettingsCollection.cs]: The `PostBack.Handlers` collection is done the same way - I don't remember why we did it like this, maybe there were some problems with generics in the VS extension in early stages of development. However, we can simplify both of them.

File: src/DotVVM.Framework/Controls/UpdateProgress.cs [hunk 3]
```diff
@@ -48,8 +70,47 @@ protected override void AddAttributesToRender(IHtmlWriter writer, IDotvvmRequest
...
+            if (!string.IsNullOrEmpty(excludedQueues) && !ValidateQueueNames(excludedQueues))
+            {
+                yield return new ControlUsageError("The ExcludedQueues must contain comma-separated list of queue names (which can contain alphanumeric characters, underscore or dash)!");
+            }
+        }
+
+        private static bool ValidateQueueNames(string queues)
+        {
+            var parts = queues.Split(',').Select(p => p.Trim());
+            return parts.All(p => Regex.IsMatch(p, @"^[a-zA-Z_][a-zA-Z0-9-_]*$"));
```
[the remark this answers is not in the document]
Comment (tomasherceg) [reply src/DotVVM.Framework/Controls/UpdateProgress.cs]: I think we don't validate the queue names on other places - we should. What regex do you suggest?

File: src/DotVVM.Framework/Resources/Scripts/DotVVM.ts (hunk 1 above)
Comment (tomasherceg) [reply src/DotVVM.Framework/Resources/Scripts/DotVVM.ts]: `noRunning` is number of running postbacks in that queue. I'll add `> 0` so it is more clear.

File: src/DotVVM.Framework/Resources/Scripts/DotVVM.ts [hunk 4]
```diff
@@ -1226,6 +1229,12 @@ class DotVVM {
             init(element: any, valueAccessor: () => any, allBindingsAccessor: KnockoutAllBindingsAccessor, viewModel: any, bindingContext: KnockoutBindingContext) {
                 element.style.display = "none";
                 var delay = element.getAttribute("data-delay");
+
+                var includedQueues = element.getAttribute("data-included-queues");
+                includedQueues = includedQueues ? includedQueues.split(",").map(q => q.trim()) : [];
```
[the remark this answers is not in the document]
Comment (exyi) [reply src/DotVVM.Framework/Resources/Scripts/DotVVM.ts]: :) I can only recommend a (free) book "You don't know JS".

File: src/DotVVM.Framework/Controls/UpdateProgress.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomasherceg) [reply src/DotVVM.Framework/Controls/UpdateProgress.cs]: I have moved the name checking method to `NamingUtils` - we can add more methods like this in that class.

File: src/DotVVM.Framework/Controls/UpdateProgress.cs [hunk 5]
```diff
@@ -23,21 +28,38 @@ public int Delay
...
         public static readonly DotvvmProperty DelayProperty =
             DotvvmProperty.Register<int, UpdateProgress>(t => t.Delay, 0);
 
-        public UpdateProgress() : base("div")
+        /// <summary>
+        /// Gets or sets the comma-separated names of PostBack.ConcurrencyQueue names for which this control should be enabled.
+        /// If not set, all queues are included automatically.
+        /// </summary>
+        [MarkupOptions(AllowBinding = false)]
+        public string IncludedQueues
```
[the remark this answers is not in the document]
Comment (tomasherceg) [reply src/DotVVM.Framework/Controls/UpdateProgress.cs]: The string arrays worked nicely and the strings are trimmed automatically - nice!

File: src/DotVVM.Framework/Resources/Scripts/DotVVM.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomasherceg) [reply src/DotVVM.Framework/Resources/Scripts/DotVVM.ts]: I did my best - hope it's a little better. But we should review the entire client-side part - there will be tons of things like that.

**Pull request merged**